server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchIgnoredMetadataPhase.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.fetch.subphase;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.index.mapper.IgnoredFieldMapper;
+import org.elasticsearch.search.fetch.FetchContext;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
+import org.elasticsearch.search.fetch.StoredFieldsSpec;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class FetchIgnoredMetadataPhase implements FetchSubPhase {
+    @Override
+    public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) throws IOException {
+        StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+        if (storedFieldsContext == null || storedFieldsContext.fetchFields() == false) {
+            return null;
+        }
+        return new FetchSubPhaseProcessor() {
+
+            SortedSetDocValues ignoredFields = null;
+
+            @Override
+            public void setNextReader(LeafReaderContext readerContext) throws IOException {
+                ignoredFields = readerContext.reader().getSortedSetDocValues(IgnoredFieldMapper.NAME);
```

Comment:
I wonder if we should make the retrieval go through value fetcher instead of directly to doc_values. One possibility is to add retrieval of metadata fields to the existing FetchFieldsPhase. This could help also delegating the version based logic and centralizing it, because there is already one or the other field type available based on the index created version. Here is my proposal, to summarize:

- remove retrieval of the stored field from StoredFieldPhase, unconditionally
- add retrieval of the _ignored meta field to FetchFieldsPhase (the phase will have to run unconditionally then and set meta fields and not only document fields)
- make the retrieval go through value fetcher
- version based logic in a single place (the existing getInstance)

Does this make sense? Am I missing anything?

Reply:
Do you mean the retrieval of all metafields?